### ISSUE TITLE
chore: set `noPropertyAccessFromIndexSignature` to true in tsconfig (on `next`-branch)

### ISSUE THREAD

**Issue by jeremyhofer** [trpc/trpc#5175]
### Provide environment information

```
  System:
    OS: Linux 6.6 Arch Linux
    CPU: (16) x64 AMD Ryzen 7 7840U w/ Radeon  780M Graphics
    Memory: 19.53 GB / 30.55 GB
    Container: Yes
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/.nvm/versions/node/v20.10.0/bin/node
    Yarn: 1.22.21 - ~/.nvm/versions/node/v20.10.0/bin/yarn
    npm: 10.2.3 - ~/.nvm/versions/node/v20.10.0/bin/npm
    pnpm: 8.12.0 - ~/.nvm/versions/node/v20.10.0/bin/pnpm
  Browsers:
    Brave Browser: 120.1.61.101
  npmPackages:
    @trpc/client: ^10.25.0 => 10.44.1
    @trpc/server: ^10.25.0 => 10.44.1
    typescript: 5.2.2 => 5.2.2
```

### Describe the bug

This is a bug related to this PR: https://github.com/trpc/trpc/pull/4796

In projects using `trpc` that are performing full type checking of dependencies, even those in node_modules, if the project has the `noPropertyAccessFromIndexSignature` typescript `compilerOption` set to `true` `TRPCClientError.ts` fails to pass type checks related to index signature vs. property access. Information on the option from typescript: https://www.typescriptlang.org/tsconfig#noPropertyAccessFromIndexSignature

Console error:
```
node_modules/@trpc/client/src/TRPCClientError.ts:42:18 - error TS4111: Property 'error' comes from an index signature, so it must be accessed with ['error'].

  40 |   return (
  41 |     isObject(obj) &&
> 42 |     isObject(obj.error) &&
     |                  ^
  43 |     typeof obj.error.code === 'number' &&
  44 |     typeof obj.error.message === 'string'
  45 |   );

node_modules/@trpc/client/src/TRPCClientError.ts:43:16 - error TS4111: Property 'error' comes from an index signature, so it must be accessed with ['error'].

  41 |     isObject(obj) &&
  42 |     isObject(obj.error) &&
> 43 |     typeof obj.error.code === 'number' &&
     |                ^
  44 |     typeof obj.error.message === 'string'
  45 |   );
  46 | }

node_modules/@trpc/client/src/TRPCClientError.ts:43:22 - error TS4111: Property 'code' comes from an index signature, so it must be accessed with ['code'].

  41 |     isObject(obj) &&
  42 |     isObject(obj.error) &&
> 43 |     typeof obj.error.code === 'number' &&
     |                      ^
  44 |     typeof obj.error.message === 'string'
  45 |   );
  46 | }

node_modules/@trpc/client/src/TRPCClientError.ts:44:16 - error TS4111: Property 'error' comes from an index signature, so it must be accessed with ['error'].

  42 |     isObject(obj.error) &&
  43 |     typeof obj.error.code === 'number' &&
> 44 |     typeof obj.error.message === 'string'
     |                ^
  45 |   );
  46 | }
  47 |

node_modules/@trpc/client/src/TRPCClientError.ts:44:22 - error TS4111: Property 'message' comes from an index signature, so it must be accessed with ['message'].

  42 |     isObject(obj.error) &&
  43 |     typeof obj.error.code === 'number' &&
> 44 |     typeof obj.error.message === 'string'
     |                      ^
  45 |   );
  46 | }
  47 |
```

### Link to reproduction

https://github.com/jeremyhofer/trpc-typescript-version-compiler-issues

### To reproduce

The bug can be reproduced by adding the `noPropertyAccessFromIndexSignature` `compilerOption` and setting it to `true`.

I was having trouble reproducing in the example stackblitz environment provided, probably due to configuration and build differences between that example and how my and other projects are configured. However, the errors are reproducible in the trpc repository itself. I have linked a fork with a commit changing the typescript version to `5.1.6` (change made due to the related issue #5176) and setting `noPropertyAccessFromIndexSignature` to `true` in `tsconfig.build.json`. Changing the compiler flag reproduces the bug in the trpc build in a number of locations when attempting to `pnpm build` after a fresh `pnpm i`.

### Additional information

Edit: relates to #5176 

Related issue from AnalogJS where myself and others have been experiencing this issue: https://github.com/analogjs/analog/issues/722

Also confirmed that downgrading `@trpc/client` to `10.38.2`, the version before https://github.com/trpc/trpc/pull/4796 was released, resolves both problems as the code in question is not present in that version.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5175">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5175/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/5175/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
Please do the two different things in separate PRs, we might not want to fix the noProp on `main` and rather on `next` as it might generate annoying conflicts where the simple workaround (and generally best practice) of skipLibCheck is simple to do

**Comment by KATT** [comment]
I'd appreciate it if you split the issues into two as well

**Comment by jeremyhofer** [comment]
:+1: sure, can split the issue. Originally grouped as both issues were introduced / related to the same PR. Will submit the other issue and update this one to remove the duplication. Will comment and ask follow up questions on proposed fixes after that is done.

**Comment by jeremyhofer** [comment]
Split complete. This issue focuses on the `noPropertyAccessFromIndexSignature` compiler option issue. #5176 focuses on the typescript 5.1.3 type narrowing issue.

**Comment by jeremyhofer** [comment]
Additional context regarding `skipLibCheck` not having an effect on these issues: https://github.com/analogjs/analog/issues/722#issuecomment-1859293284

**Comment by KATT** [comment]
This setting should be turned on and fixed with `next` as a base, I don't care enough about to fix conflicts since I think `skipLibCheck: true` is a sane default anyway.

**Comment by jeremyhofer** [comment]
This is addressed in #5179
